chore(curriculum): add flexbox transcripts (#58929)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lecture-working-with-css-flexbox/672bd658c0e190f674a5e057.md

# --transcript--

What are some common flex properties, and how do they work?

`flex` properties are properties that you can apply to `flex` containers to determine the distribution of child elements. We’ll cover some of the most commonly used ones: `flex-wrap`, `justify-content`, and `align-items`.

Let’s start with `flex-wrap`. This property determines how `flex` items are wrapped within a `flex` container to fit the available space. `flex-wrap` can take three possible values: `nowrap`, `wrap`, and `wrap-reverse`. `nowrap` is the default value—`flex` items won’t be wrapped onto a new line, even if their `width` exceeds the container's `width`.

In the code below, we have three `div` elements. Let’s focus on the `width`. The `main` container with the `red` `border` has a `width` of `200px`, while its three child `div` elements combined have a `width` of `240px` (`80px` each):

```css
main {
  width: 200px;
  display: flex;
  border: 2px solid red;
}

div {
  width: 80px;
  height: 50px;
}

#first-div {
  background-color: #4d70b2;
}

#second-div {
  background-color: #5c4db2;
}

#third-div {
  background-color: #4da3b2;
}
```

> The `width` of the `div` elements exceeds the `width` of their container, but by default they will be shrunk to fit the available space. If you do want to wrap them when they exceed the `width` of their container, you can set `flex-wrap: wrap` on the `flex` container:

Reviewer: ```suggestion
The width of the `div` elements exceeds the width of their container, but by default they will be shrunk to fit the available space. If you do want to wrap them when they exceed the width of their container, you can set `flex-wrap: wrap` on the flex container:
```

```css
flex-wrap: wrap;
```

The `div` elements will be rearranged in rows when they exceed the `width` of their container. You can wrap `flex` items in reverse order with `flex-wrap: wrap-reverse`:

```css
flex-wrap: wrap-reverse;
```

The `flex-flow` property is a shorthand property for `flex-direction` and `flex-wrap`. In this example, we set `flex-direction` to `column` and `flex-wrap` to `wrap-reverse`:

```css
flex-flow: column wrap-reverse;
```

Great. Now let’s talk about `justify-content`. `justify-content` aligns the child elements along the main axis of the `flex` container. If you assign the value `flex-start` to `justify-content`, the `flex` items will be aligned to the start of the main axis. This could be horizontal or vertical:

```css
justify-content: flex-start;
```

With `justify-content: flex-end`, `flex` items are aligned to the end of the main axis, horizontally or vertically:

```css
justify-content: flex-end;
```

> To center the `flex` items along the main axis, you can use `justify-content: center`:

Reviewer: ```suggestion
To center the flex items along the main axis, you can use `justify-content: center`:
```

```css
justify-content: center;
```

> To distribute the elements evenly along the main axis, you can use `justify-content: space-between`. This will add some space between the `flex` items if needed:

Reviewer: ```suggestion
To distribute the elements evenly along the main axis, you can use `justify-content: space-between`. This will add some space between the flex items if needed:
```

```css
justify-content: space-between;
```

> `justify-content: space-around` distributes `flex` items evenly within the main axis, adding a space before the first item and after the last item. This additional space is half of the space between each pair of adjacent items. If there’s only one item to distribute, it will be centered:

Reviewer: ```suggestion
`justify-content: space-around` distributes flex items evenly within the main axis, adding a space before the first item and after the last item. This additional space is half of the space between each pair of adjacent items. If there's only one item to distribute, it will be centered:
```

```css
justify-content: space-around;
```

And last but not least, we have `justify-content: space-evenly`, which distributes the items evenly along the main axis. The space between the items and the space before and after the first and last elements, are exactly the same:

```css
justify-content: space-evenly;
```

> Great. Now you know how to distribute `flex` items along the main axis. But you may also want to distribute them along the cross axis. Remember that the cross axis is perpendicular to the main axis. You can do this with the `align-items` property. To center the items along the cross axis, you just need to add `align-items: center` to the `flex` container: 

Reviewer: ```suggestion
Great. Now you know how to distribute flex items along the main axis. But you may also want to distribute them along the cross axis. Remember that the cross axis is perpendicular to the main axis. You can do this with the `align-items` property. To center the items along the cross axis, you just need to add `align-items: center` to the flex container: 
```

```css
align-items: center;
```

> In this example, the `flex` items are centered along the cross axis, which is vertical by default. If the cross axis is horizontal, they will be centered horizontally instead. In contrast, `align-items: flex-start` aligns the items to the start of the cross axis:

Reviewer: ```suggestion
In this example, the flex items are centered along the cross axis, which is vertical by default. If the cross axis is horizontal, they will be centered horizontally instead. In contrast, `align-items: flex-start` aligns the items to the start of the cross axis:
```

```css
align-items: flex-start;
```

> The opposite is `justify-content: flex-end`. This will align `flex` items to the end of the cross axis, vertically or horizontally:

Reviewer: ```suggestion
The opposite is `justify-content: flex-end`. This will align flex items to the end of the cross axis, vertically or horizontally:
```

```css
justify-content: flex-end;
```

> To stretch the `flex` items along the cross axis, you can use `align-items: stretch`:

Reviewer: ```suggestion
To stretch the flex items along the cross axis, you can use `align-items: stretch`:
```

```css
align-items: stretch;
```

> This also works with elements that are automatically sized, such as those without set `width` or `height`, or with only a minimum `width` or `height`. The `flex` items will stretch to fill the container in the direction of the cross axis.

Reviewer: ```suggestion
This also works with elements that are automatically sized, such as those without set `width` or `height`, or with only a minimum `width` or `height`. The flex items will stretch to fill the container in the direction of the cross axis.
```

> 

Reviewer: ```suggestion

And finally, you can use the `align-self` property to assign a different alignment on the cross axis to an individual flex item.

For example, you can stretch it with `align-self: stretch`. You can center it with `align-self: center`. You can align it to the start of the cross axis with `align-self: flex-start`. Or you can align it to the end of the cross axis with `align-self: flex-end`.

```
This reflects the changes made in the updated video.

> There are other `flex` properties and values that you can choose from to create the responsive layout that you envision, but these are the most commonly-used ones. With these CSS `flex` properties and your new knowledge of the CSS `flex` model, you can start creating responsive layouts to create a smooth and inclusive user experience across devices.

Reviewer: ```suggestion
There are other flex properties and values that you can choose from to create the responsive layout that you envision, but these are the most commonly-used ones. With these CSS flex properties and your new knowledge of the CSS flex model, you can start creating responsive layouts to create a smooth and inclusive user experience across devices.
```
